util/rate_limiter.cc

```diff
@@ -275,52 +279,42 @@ int64_t GenericRateLimiter::CalculateRefillBytesPerPeriod(
 }
 
 Status GenericRateLimiter::Tune() {
-  const int kLowWatermarkPct = 50;
-  const int kHighWatermarkPct = 90;
-  const int kAdjustFactorPct = 5;
-  // computed rate limit will be in
-  // `[max_bytes_per_sec_ / kAllowedRangeFactor, max_bytes_per_sec_]`.
   const int kAllowedRangeFactor = 20;
+  const int kRatioLower = 15;
+  const int kRatioUpper = 50;
+  const int kRatioPadding = 5;
 
   std::chrono::microseconds prev_tuned_time = tuned_time_;
   tuned_time_ = std::chrono::microseconds(NowMicrosMonotonic(env_));
-
-  int64_t elapsed_intervals = (tuned_time_ - prev_tuned_time +
-                               std::chrono::microseconds(refill_period_us_) -
-                               std::chrono::microseconds(1)) /
-                              std::chrono::microseconds(refill_period_us_);
-  // We tune every kRefillsPerTune intervals, so the overflow and division-by-
-  // zero conditions should never happen.
-  assert(num_drains_ - prev_num_drains_ <= port::kMaxInt64 / 100);
-  assert(elapsed_intervals > 0);
-  int64_t drained_pct =
-      (num_drains_ - prev_num_drains_) * 100 / elapsed_intervals;
+  auto duration = tuned_time_ - prev_tuned_time;
+  auto duration_ms =
+      std::chrono::duration_cast<std::chrono::milliseconds>(duration).count();
 
   int64_t prev_bytes_per_sec = GetBytesPerSecond();
-  int64_t new_bytes_per_sec;
-  if (drained_pct == 0) {
-    new_bytes_per_sec = max_bytes_per_sec_ / kAllowedRangeFactor;
-  } else if (drained_pct < kLowWatermarkPct) {
-    // sanitize to prevent overflow
-    int64_t sanitized_prev_bytes_per_sec =
-        std::min(prev_bytes_per_sec, port::kMaxInt64 / 100);
-    new_bytes_per_sec =
-        std::max(max_bytes_per_sec_ / kAllowedRangeFactor,
-                 sanitized_prev_bytes_per_sec * 100 / (100 + kAdjustFactorPct));
-  } else if (drained_pct > kHighWatermarkPct) {
-    // sanitize to prevent overflow
-    int64_t sanitized_prev_bytes_per_sec = std::min(
-        prev_bytes_per_sec, port::kMaxInt64 / (100 + kAdjustFactorPct));
-    new_bytes_per_sec =
-        std::min(max_bytes_per_sec_,
-                 sanitized_prev_bytes_per_sec * (100 + kAdjustFactorPct) / 100);
-  } else {
-    new_bytes_per_sec = prev_bytes_per_sec;
+
+  bytes_sampler_.AddSample(duration_bytes_through_ * 1000 / duration_ms);
+  highpri_bytes_sampler_.AddSample(duration_highpri_bytes_through_ * 1000 /
+                                   duration_ms);
+  duration_bytes_through_ = 0;
+  duration_highpri_bytes_through_ = 0;
+  auto util = bytes_sampler_.GetRecentValue() * 100 / prev_bytes_per_sec;
+  if (util >= 98) {
```

Comment:
add inline comment for the purpose of `ratio_delta`? looks to me it is to boost the ratio adjustment.